[WIP] Redgate Laserdome

KillianKirilenko opened the pull request:
I've been contemplating whether or not to kick this one off for a couple weeks, so here it actually is - THE LASERDOME! Ten sophonts enter, ten sophonts leave! It's ~~violent~~ cathartic fun for the whole family!

Upon entering the redgate, you'll be dropped in the communal 'safe zone' where you can grab a vest and gun for your team. From there, it's up to players to fan out and protect their team's flag whilst attempting to capture the enemy flag. If you successfully bring the enemy flag to your base and use it on the flag stand, your team scores! The enemy flag is then teleported back to their base. But beware; carrying the flag slows you down a bit, and it can't be stowed in bags. Protect your team's flag carrier, and intercept the enemy's!

![laserdome_wip](https://github.com/VOREStation/VOREStation/assets/49700375/c11fcd10-368c-453c-9428-730921239ba1)
![image](https://cdn.discordapp.com/attachments/1138986141190729820/1156108786658988082/image.png)
![image](https://cdn.discordapp.com/attachments/1138986141190729820/1156122260709584978/image.png)

To-do:
- [x] The base map!
- [x] Capture The Flag functionality.
- [ ] Limit the announcements to the laserdome z-level and/or entertainment frequency
- [ ] Announcements for when a flag is dropped?
- [ ] Mapping refinements

This isn't your standard redgate map in that it has no voremobs thusfar, but things are still being worked on. It could potentially be made a sub/extra level of the city map, but we'll see how things pan out. It also only uses maybe half the total z-level space, so in future it could be revized to have two seperate arenas - one free-for-all and one capture-the-flag, or perhaps a different game mode!

SatinIsle commented:
Really cool idea! Arena that you've got atm looks really cool to me.

You've got a lot of empty space on the map, of course, and you wouldn't want to ruin this arena by making it too big. Perhaps you could have another two separate arenas, maybe accessible via portals from the centre, to pack a bit more variety into the map? Could be a place for vore mobs if you want to add them too, maybe one of those arenas uses them as a hazard or something. None of this needs to be done or anything, just throwing ideas out there.

KillianKirilenko commented:
> Really cool idea! Arena that you've got atm looks really cool to me.
> 
> You've got a lot of empty space on the map, of course, and you wouldn't want to ruin this arena by making it too big. Perhaps you could have another two separate arenas, maybe accessible via portals from the centre, to pack a bit more variety into the map? Could be a place for vore mobs if you want to add them too, maybe one of those arenas uses them as a hazard or something. None of this needs to be done or anything, just throwing ideas out there.

I direct your attention to the closing paragraph.

SatinIsle commented:
> > Really cool idea! Arena that you've got atm looks really cool to me.
> > You've got a lot of empty space on the map, of course, and you wouldn't want to ruin this arena by making it too big. Perhaps you could have another two separate arenas, maybe accessible via portals from the centre, to pack a bit more variety into the map? Could be a place for vore mobs if you want to add them too, maybe one of those arenas uses them as a hazard or something. None of this needs to be done or anything, just throwing ideas out there.
> 
> I direct your attention to the closing paragraph.
> 
> (Which is to say; yes, I have considered these things already, but will worry about them later.)

Sorry, somehow I completely blanked on that one line after testing the map out!
Thinking of another game mode is an interesting idea too, free-for-all is obvious, but a third one is tougher. I know you're not worrying about it for now, but just for the sake of throwing ideas out there, albeit they might be a bit code heavy: Capture the point type mode. Or maybe even a more cooperative "wave defence" mode in which the group has to stop some mobs reaching them? Possibly dreaming a bit big on that/out of scope.

KillianKirilenko commented:
What I'll probably end up doing for a second team-v-team mode is a "football" type game mode; take a neutral item from the center and dunk it in the enemy team's goal to score a point. The item then returns to the center field automatically, similar to how scoring flags works currently.

On that note, I've made it so that touching your team's flag will instantly teleport it back to the flag base instead of it needing to be carried back. I might change this to a channeling/do_after mechanic, or have flags teleport back to their base if they're left unattended for too long. I also want to make it so they can't be taken off the zlevel at some point, but that probably won't be too hard. Probably.

(Reminds me, I also wanted to make a dodgeball...)

KillianKirilenko commented:
Small update; added some redgate functionality to block transit whilst you're carrying a prohibited item.

KillianKirilenko commented:
Now working on the hyperball arena. Need to do some step triggers/teleporters/whatever-ya-call'ems for the new hub that takes players into spawn-like zones behind the scoring areas. There's now a main concourse area that I'm working on as well.

KillianKirilenko commented:
OK, we are now *much* more presentable and theoretically in a deployable state! The concourse now has a fast food restaurant, first aid station, a couple of empty storefronts, and some showers for players to "cool off" post game.

There are also scads of entertainment encryption keys scattered around the concourse so people can listen in on ongoing matches.

KillianKirilenko commented:
Some more layout polish to fix portal oddities, a bit of general shine on the concourse (such as mannequins in the display cases), and inhand sprites for the ball/flag to make it obvious who the carrier is! I'd love to have some fancier animated stuff for the hyperball carrier, but the rings will do for now because I'm no good at detailed animated effects myself.

All of the team gear lockers have been provided with a pair of size-standardization bracelets so over/under sized players without access to sizeguns or NIF sizeshifting can participate with less concerns over size (dis)advantages. The hyperball goals have also been polished up a bit so that the energy net is animated and the voids are filled with `1,1,1,1` pixels so you can dunk/throw more easily. Likewise the flag base background has been filled with `1,1,1,1` pixels to make scoring easier. Finally, the hyperball now turns the colour of the team that touched it when picked up and resets to green after someone scores.

KillianKirilenko commented:
Added some extra functionality I'd been meaning to figure out, namely: multi-hit rules! Lasertag guns can now be toggled between multi-hit and "instagib" rules; they get twice as many shots (so 20 total) and fire much faster in multi-hit rules, but you need to hit someone three* times in order to drop them. Instagib rules are the classic "one hit and you're out" behaviour with a slower firerate, and it's what they default to.

*A new define in the lasertag vests file can increase or decrease this limit.

I have also added some laserdome-specific equipment lockers so I don't need to edit those constantly and worry about keeping them in synch (and also so jumpsuits and helmets are included).

Heroman3003 commented:
Oh I kept seeing 14/18 tasks and never looked to see that last 4 are optional
